File: 990-verifying-an-alien-dictionary/verifying-an-alien-dictionary.py
class Solution:
    def isAlienSorted(self, words, order):
        rank = {}

        for i in range(len(order)):
            rank[order[i]] = i

        for i in range(len(words) - 1):

            first = words[i]
            second = words[i + 1]

            same = True

            for j in range(min(len(first), len(second))):

                if first[j] != second[j]:

                    same = False

                    if rank[first[j]] > rank[second[j]]:
                        return False

                    break

            if same and len(first) > len(second):
                return False

        return True

File: 990-verifying-an-alien-dictionary/verifying-an-alien-dictionary.java
class Solution {
    public boolean isAlienSorted(String[] words, String order) {

        int[] rank = new int[26];

        for (int i = 0; i < order.length(); i++) {
            rank[order.charAt(i) - 'a'] = i;
        }

        for (int i = 0; i < words.length - 1; i++) {

            String first = words[i];
            String second = words[i + 1];

            boolean same = true;

            int length = Math.min(
                first.length(),
                second.length()
            );

            for (int j = 0; j < length; j++) {

                if (first.charAt(j) != second.charAt(j)) {

                    same = false;

                    if (
                        rank[first.charAt(j) - 'a'] >
                        rank[second.charAt(j) - 'a']
                    ) {
                        return false;
                    }

                    break;
                }
            }

            if (same && first.length() > second.length()) {
                return false;
            }
        }

        return true;
    }
}